modules/AlphaWalletFoundation/AlphaWalletFoundation/Transfer/eip1559/SuggestEip1559.swift

```diff
@@ -0,0 +1,220 @@
+                    block: .blockNumber(value: lastBlock),
+                    rewardPercentile: rewardPercentile).first
+
+                let rewardSize = feeHistory.reward.count
+                guard !feeHistory.reward.isEmpty else { throw EIP1559Error.rewardsNotFound }
+                print("[GasEstimator] rewards raw: \(feeHistory.reward)")
+                for reward in feeHistory.reward {
+                    let value = reward[0]
+                    rewards += [value]
+                }
+                if rewardSize < blockCount {
+                    break
+                }
+                needBlocks -= blockCount
+            }
+            ptr -= blockCount + 1
+        }
+
+        if rewards.isEmpty {
+            return fallbackPriorityFee
+        }
+        rewards.sort()
+
+        let index = Int(floor(Double((rewards.count - 1) * rewardBlockPercentile) / 100.0))
+
+        print("[GasEstimator] rewards: \(rewards)")
+        print("[GasEstimator] priorityFee: \(rewards[index])")
+        print("[GasEstimator] priorityFee index: \(index)")
+
+        return Decimal(rewards[index])
+    }
+
+    enum EIP1559Error: Error {
+        case baseFeeIsEmpty
+        case firstBlockMissing
+        case failureToBuildTimedEstimate
+        case indexedRewardNotFound
+        case rewardsNotFound
+        case indexedBaseFeeNotFound
+        case minBaseFeeNotFound
+    }
+
+    func calculateResult(priorityFee: Decimal, feeHistory: FeeHistory) throws -> [Int: EIP1559FeeOracleResult] {
+        var baseFee = feeHistory.baseFeePerGas
+
+        print("[GasEstimator] baseFee.-1: \(feeHistory.baseFeePerGas)")
+        print("[GasEstimator] baseFee.0: \(baseFee)")
+
+        guard !baseFee.isEmpty else { throw EIP1559Error.baseFeeIsEmpty }
+
+        print("[GasEstimator] last in base fee.0: \(baseFee[baseFee.count - 1])")
+
+        let lastBaseFee = Double(baseFee[baseFee.count - 1]) * 1.125
+        baseFee[baseFee.count - 1] = lastBaseFee
+        print("[GasEstimator] last value: \(lastBaseFee)")
+
+        for i in feeHistory.gasUsedRatio.indices.reversed() where feeHistory.gasUsedRatio[i] > 0.9 {
+            baseFee[i] = baseFee[i + 1]
+        }
+
+        let order: [Int] = Array(0 ... feeHistory.gasUsedRatio.count).indices.sorted {
+            guard let first = baseFee[safe: $0] else { return false }
+            guard let second = baseFee[safe: $1] else { return true }
+            return first < second
+        }
+
+        print("[GasEstimator] order: \(order)")
+
+        var maxBaseFee: Decimal = .zero
+        var result: [Int: EIP1559FeeOracleResult] = [:]
+
+        print("[GasEstimator] baseFee.1: \(baseFee)")
+        print("[GasEstimator] timeFactors: \(Array(Array(0 ..< maxTimeFactor).reversed()))")
+
+        for timeFactor in Array(Array(0 ..< maxTimeFactor).reversed()) {
+            var bf = try predictMinBaseFee(baseFee: baseFee, order: order, timeFactor: timeFactor)
```

Comment:
It feels like there might be a flaw in the computation for `predictMinBaseFee()`. As you go down the list and make calls to it, it generates values in a curve that isn't always increasing (or decreasing, depending on which direction you look at it), so that makes the final estimates not strictly increasing/decreasing too. Maybe it's because the gas usage goes up and down and it doesn't take that into account correctly?

Barring studying and fixing the algorithm (or our implementation of it, which seems to follow the Kotlin version linked to), I think a simple patch is to just store the value of `bf` in each loop and make sure it doesn't go up (or go down, I didn't check which direction)

```
var bf = try predictMinBaseFee(baseFee: baseFee, order: order, timeFactor: timeFactor)
```

eg. 

```
var largestBf = 0 //TODO: need to check if largest or smallest and a sensible default or nil
for timeFactor in (0...maxTimeFactor).reversed() {
    var bf = try predictMinBaseFee(baseFee: baseFee, order: order, timeFactor: timeFactor)
    bf = max(bf, largestBf) //TODO or figure if it should be: min(bf, largestBf)
```